src/axom/primal/operators/winding_number.hpp

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file winding_number.hpp
+ *
+ * \brief Consists of methods to compute winding numbers for points 
+ *        with respect to various geometric objects.
+ */
+
+#ifndef AXOM_PRIMAL_WINDING_NUMBER_HPP_
+#define AXOM_PRIMAL_WINDING_NUMBER_HPP_
+
+// Axom includes
+#include "axom/config.hpp"
+
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Segment.hpp"
+#include "axom/primal/geometry/Triangle.hpp"
+#include "axom/primal/geometry/Polygon.hpp"
+#include "axom/primal/geometry/Polyhedron.hpp"
+#include "axom/primal/geometry/BezierCurve.hpp"
+#include "axom/primal/geometry/CurvedPolygon.hpp"
+#include "axom/primal/operators/detail/winding_number_impl.hpp"
+
+// C++ includes
+#include <cmath>
+
+namespace axom
+{
+namespace primal
+{
+/*
+ * \brief Compute the winding number with respect to a line segment
+ *
+ * \param [in] q The query point to test
+ * \param [in] s The line segment
+ * \param [in] edge_tol The tolerance at which a point is on the line
+ *
+ * \return double The generalized winding number
+ */
+template <typename T>
+double winding_number(const Point<T, 2>& q,
+                      const Segment<T, 2>& s,
+                      double edge_tol = 1e-8)
+{
+  return detail::linear_winding_number(q, s[0], s[1], edge_tol);
+}
+
+/*
+ * \brief Compute the winding number with respect to a 2D triangle
+ *
+ * \param [in] q The query point to test
+ * \param [in] tri The triangle
+ * \param [in] edge_tol The tolerance at which a point is on the line
+ *
+ * \return int The integer winding number
+ */
+template <typename T>
+int winding_number(const Point<T, 2>& q,
+                   const Triangle<T, 2>& tri,
+                   bool useStrictInclusion = false,
+                   double edge_tol = 1e-8)
+{
+  return winding_number(
+    q,
+    Polygon<T, 2>(axom::Array<Point<T, 2>>({tri[0], tri[1], tri[2]})),
+    useStrictInclusion,
+    edge_tol);
```

Comment:
This is the best optimization I can think of. While it's possible to construct the winding number for the triangle as a sum of linear winding numbers for each edge, it's easiest to just check containment + orientation as a polygon. The `winding_number( Polygon )` overload is, I believe anyway, to be the most efficient way of doing this, as it can (usually) solve the problem just with floating point comparisons. 

There actually is a `Triangle::checkInTriangle( Point )` method that accomplishes the same thing, but I believe that method is less efficient, since it converts the query to barycentric coordinates first, which it computes directly as ratios of areas. Even in the best case, that involves more floating point multiplications.